index.bs

```diff
@@ -2793,25 +2793,30 @@ Note: Invoking this method from a [=browsing context=] where the [=Web Authentic
 
 </div>
 
-### Availability of a [=passkey platform authenticator=] - PublicKeyCredential's `isPasskeyPlatformAuthenticatorAvailable()` Method ### {#sctn-isPasskeyPlatformAuthenticatorAvailable}
+### Availability of [=client=] capabilities - PublicKeyCredential's `getClientCapabilities()` Method ### {#sctn-getClientCapabilities}
 
-<div link-for-hint="WebAuthentication/isPasskeyPlatformAuthenticatorAvailable">
+<div link-for-hint="WebAuthentication/getClientCapabilities">
 
-[=[WRPS]=] use this method to determine whether they can create a new [=passkey=] using a [=user-verifying platform authenticator=] or a {{AuthenticatorTransport/hybrid}} authenticator.
-Upon invocation, the [=client=] employs a [=client platform=]-specific procedure to discover available [=user-verifying platform authenticators=] and the
-availability of {{AuthenticatorTransport/hybrid}} transport.
-If one or both are discovered, the promise is resolved with the value of [TRUE].
-If neither is discovered, the promise is resolved with the value of [FALSE].
-Based on the result, the [=[RP]=] can take further actions to guide the user to create a [=passkey=].
+[=[WRPS]=] use this method to determine the availability of a limited set of [=WebAuthn Client|client=] capabilities to offer certain workflows and experiences to users. For example, an RP may offer a sign in button on clients where only {{AuthenticatorTransport/hybrid}} transport is available or where {{CredentialMediationRequirement/conditional}} mediation is unavailable (instead of showing a username field).
 
-This method has no arguments and returns a Boolean value.
+Upon invocation, the [=client=] employs a [=client platform=]-specific procedure to discover availablity of these capabilities.
+
+This method has no arguments and returns a record of capability keys to Boolean values.
 
 <xmp class="idl">
     partial interface PublicKeyCredential {
-        static Promise<boolean> isPasskeyPlatformAuthenticatorAvailable();
+        static Promise<PublicKeyCredentialClientCapabilities> getClientCapabilities();
     };
+
+    typedef record<DOMString, boolean> PublicKeyCredentialClientCapabilities;
 </xmp>
 
+The set of [=map/keys=] SHOULD equal the set of [=enumeration values=] of {{ClientCapability}}.
+
+When the value for a given capability is [TRUE], the feature is known to be currently supported by the client.
+When the value for a given capability is [FALSE], the feature is known to be not currently supported by the client.
+When a capability does not [=map/exist=] as a key, the availability of the client feature is not known.
```

Comment:
We should link to `{#sctn-disclosing-client-capabilities}` somewhere, maybe here?

```suggestion
When a capability does not [=map/exist=] as a key, the availability of the client feature is not known.
See also [[#sctn-disclosing-client-capabilities]].
```

Reply:
Oh, you know what, I think I found a good place for that link to the Disclosing Client Capabilities section, how about this?

![Screenshot 2024-01-08 at 9 48 33 AM](https://github.com/w3c/webauthn/assets/5166470/f853962d-3769-4cf6-bfdc-7713af3847b8)

Reply:
Hmyeah, good point - right now it is a bit ambiguous which of "The set of keys SHOULD equal the set of enumeration values" and "Client implementations MAY wish to limit capability disclosures" takes precedence. I don't think `null` or `undefined` are legal values of a `record<DOMString, boolean>`, so we'd need to tweak the type a bit in order to support your suggestion. It seems to me like both `record<DOMString, boolean | undefined>` and `record<DOMString, boolean?>` (nullable boolean) are legal, so I think we could choose either `undefined` or `null` for supported but undisclosed features.

But on the other hand, I think it's better to rephrase the "The set of keys SHOULD equal the set of enumeration values" requirement instead of complicating the type:

- `undefined` vs. `null`: A key being absent vs. its value being `undefined` are largely equivalent in JavaScript implementations (one difference being that `Object.keys({foo: undefined}) != Object.keys({})`), so between `undefined` and `null` I think we should choose the former so we don't end up with _4_ possible values.

- Key absent vs. value `undefined`: The API and its consumers needs to be ready to handle different browsers with different sets of feature flags, so there will always be the possibility that a feature flag is defined in the spec but absent from the set of keys. So you could get `{ newFeature: false, hybridTransport: true }` in one browser and `{ hybridTransport: true }` in another browser (or older version of the same browser), even if both browsers follow the recommendation to include keys for all known features.

  So then there are again 4 different possibilities: the key can be present or absent, and the value can be `false`, `true` or `undefined`. Is it useful to know that the browser knows about the existence of a feature identifier, but doesn't wish to disclose whether it's supported? (There doesn't seem to be a JS binding to obtain the list of enumeration values, but `Object.keys()` would let you iterate over the keys present.)

  Actually, I guess that too could be a fingerprinting opportunity and thus a privacy concern. I would err on the side of caution and have keys be absent if the feature is unknown or if the browser doesn't wish to disclose it. This also reduces the number of possible outputs to 3 per feature.

So, how about this? :smile:

```
The set of [=map/keys=] SHOULD equal the set of [=enumeration values=] of {{ClientCapability}},
but the client MAY omit keys as it deems necessary; see [[#sctn-disclosing-client-capabilities]].
```

Reply:
Thanks for thinking out loud here, I follow your train of thought. I agree this can turn into a real headache if we try to require defining unknown capability keys as `undefined` instead of allowing those keys to be omitted. I'll add in your MAY addition above.